systemd/manager.go

```diff
@@ -218,27 +258,49 @@ func (m *systemdUnitManager) GetUnitStates(filter pkg.Set) (map[string]*unit.Uni
 		if h, ok := m.hashes[dus.Name]; ok {
 			us.UnitHash = h.String()
 		}
-		states[dus.Name] = us
+		unitStates[dus.Name] = us
+		if m.enableUnitStateCache {
+			delete(m.unitStateCache.registerUnitOperation, dus.Name)
+		}
 	}
 
-	// grab data on subscribed units that didn't show up in ListUnits, most
-	// likely due to being inactive
+	unitStates, err = m.inactiveUnitStates(filter, unitStates)
+
+	if m.enableUnitStateCache {
+		m.unitStateCache.unitStates = unitStates
+	}
+	return unitStates, nil
+}
+
+// This function grabs data on subscribed units that didn't show up in ListUnits, most
+// likely due to being an inactive state
+func (m *systemdUnitManager) inactiveUnitStates(filter pkg.Set, unitStates map[string]*unit.UnitState) (map[string]*unit.UnitState, error) {
 	for _, name := range filter.Values() {
-		if _, ok := states[name]; ok {
+		if _, ok := unitStates[name]; ok {
 			continue
 		}
 
-		us, err := m.getUnitState(name)
-		if err != nil {
-			return nil, err
-		}
-		if h, ok := m.hashes[name]; ok {
-			us.UnitHash = h.String()
+		// If there is any operation previously registered we collect the states from systemd
+		_, ok := m.unitStateCache.registerUnitOperation[name]
+		if m.enableUnitStateCache && !ok && m.unitStateCache.unitStates[name] != nil {
+			unitStates[name] = m.unitStateCache.unitStates[name]
+		} else {
+			us, err := m.getUnitState(name)
+			if err != nil {
+				return unitStates, err
```

Comment:
Previously GetUnitStates() was returning a nil and an error here, now errors are ignored and states are returned.